R/plotreg.R

```diff
@@ -1,394 +1,372 @@
-      dataframe <- reorder(dataframe, reorder.coef)
-      co.names <- dataframe[, 1]
-      co <- dataframe[, 2]
-      lower.inner <- dataframe[, 3]
-      upper.inner <- dataframe[, 4]
-      lower.outer <- dataframe[, 5]
-      upper.outer <- dataframe[, 6]
-      
-      signif.outer <- TRUE
-      note <- "Bars denote CIs."
-      message(paste0("Model ", i, ": bars denote 0.5 (inner) resp. ", ci.level, 
-          " (outer) confidence intervals (computed from standard errors)."))
+        z.inner <- qnorm(1 - ((1 - 0.5) / 2))
+        lower.inner <- dataframe$co - (z.inner * dataframe$se)
+        upper.inner <- dataframe$co + (z.inner * dataframe$se)
+        z.outer <- qnorm(1 - ((1 - ci.level) / 2))
+        lower.outer <- dataframe$co - (z.outer * dataframe$se)
+        upper.outer <- dataframe$co + (z.outer * dataframe$se)
+        
     } else if (length(models[[i]]@se) == 0 && length(models[[i]]@pvalues) > 0) {
-      stop("Model has p-values but no SEs. SEs or CIs are required for plotting.")
+        stop("Model has p-values but no SEs. SEs or CIs are required for plotting.")
     } else {
-      co <- models[[i]]@coef
-      co.names <- models[[i]]@coef.names
-      co <- co[-remove.rows]
-      co.names <- co.names[-remove.rows]
-      lower.outer <- models[[i]]@ci.low
-      upper.outer <- models[[i]]@ci.up
-      lower.inner <- NULL
-      upper.inner <- NULL
-      lower.outer <- lower.outer[-remove.rows]
-      upper.outer <- upper.outer[-remove.rows]
-      
-      # sort according to reorder.coef argument
-      dataframe <- data.frame(co.names, co, lower.outer, upper.outer)
-      dataframe <- reorder(dataframe, reorder.coef)
-      co.names <- dataframe[, 1]
-      co <- dataframe[, 2]
-      lower.outer <- dataframe[, 3]
-      upper.outer <- dataframe[, 4]
-      
-      signif.outer <- TRUE
-      note <- "Bars denote CIs."
-      message(paste0("Model ", i, ": bars denote  ", ci.level, 
-          " confidence intervals."))
+        lower.outer <- dataframe$ci.low
+        upper.outer <- dataframe$ci.up
+        lower.inner <- rep(0, length(dataframe$co.names))
+        upper.inner <- rep(0, length(dataframe$co.names))
+    }
+    
+    if (length(lower.outer) > 0) {
+        dataframe <- cbind(dataframe, lower.outer)     
+    }
+    if (length(lower.inner) > 0) {
+        dataframe <- cbind(dataframe, lower.inner)     
+    }
+    if (length(upper.outer) > 0) {
+        dataframe <- cbind(dataframe, upper.outer)     
+    }    
+    if (length(upper.inner) > 0) {
+        dataframe <- cbind(dataframe, upper.inner)     
     }
+    if (length(dataframe$pv) > 0) {
+        signif.outer <- dataframe$pv < (1 - ci.level)
+    } 
     
-    if (!is.null(custom.note)) {
-      note <- custom.note
+    # which terms are significant?
+    if(length(dataframe$ci.low) > 0) {
+        signif.outer <- ((dataframe$ci.low) > 0 & (dataframe$ci.up) > 0 | (dataframe$ci.low) < 0 & (dataframe$ci.up) < 0)
```

Comment:
I am pretty sure there is a bug here due to the location of the parentheses. Shouldn't it look more like the following?
```
signif.outer <- ((dataframe$ci.low > 0 && dataframe$ci.up > 0) ||
                 (dataframe$ci.low < 0 && dataframe$ci.up < 0))
```
I think in my original code, there must have been a different way to establish significance; perhaps of interest in comparison.

I'm not sure what exactly you were trying to do in this line, but my suggested changes may have knock-on effects for the following lines. Please check.

Reply:
This bit of code works only when the model has a CI and not a pvalue (e.g btergm). 

It basically says that if both lower and upper CIs are below or above zero is significant. 

In your original code, it was split between coefplot and plotreg. In plotreg you were assigning significance in a different way in each piece of the for loop and then it was confronted with the part defined in coefplot. 

I changed it to two options only: if there is a pvalue and if there is a CI. I wrote it a bit more nicely now:
 ```
# # which terms are significant?
    if (length(dataframe$pv) > 0) {
        signif.outer <- dataframe$pv < (1 - ci.level)
    } else if (length(dataframe$ci.low) > 0) {
        signif.outer <- ((dataframe$ci.low) > 0 & (dataframe$ci.up) > 0 | (dataframe$ci.low) < 0 & (dataframe$ci.up) < 0)
    }

```

This way it works very well, but I'm happy to work more on this if you want me to.

Reply:
I still think the line with the four conditions does not do what you think it does. You want to group the first two and the second two together, respectively. But you don't do that. See my line above for comparison.

Reply:
```
# tergm example
set.seed(1912)

library(xergm)
networks <- list()
for(i in 1:10){            # create 10 random networks with 10 actors
    mat <- matrix(rbinom(100, 1, .25), nrow = 10, ncol = 10)
    diag(mat) <- 0           # loops are excluded
    nw <- network(mat)       # create network object
    networks[[i]] <- nw      # add network to the list
}

covariates <- list()
for (i in 1:10) {          # create 10 matrices as covariate
    mat <- matrix(rnorm(100), nrow = 10, ncol = 10)
    covariates[[i]] <- mat   # add matrix to the list
}

m1 <- btergm(networks ~ edges + istar(2) +
                         edgecov(covariates), R = 100)

plotreg(list(m1, m1, m1, m1))
```

with my line 

`signif.outer <- ((dataframe$ci.low) > 0 & (dataframe$ci.up) > 0 | (dataframe$ci.low) < 0 & (dataframe$ci.up) < 0)`, 

the  `cat `output is: 

`TRUE TRUE FALSE TRUE TRUE FALSE TRUE TRUE FALSE TRUE TRUE FALSE`

which is the significance for each covariate.

with your line 

```
signif.outer <- ((dataframe$ci.low > 0 && dataframe$ci.up > 0) ||
                 (dataframe$ci.low < 0 && dataframe$ci.up < 0)) ```

the `cat` output is 

`TRUE`

that provide significance only for the first term. 

Not sure why honestly.

Reply:
If you have tested it thoroughly, it will be fine. But it's not clear to me which conjunction of conditions takes priority in your line. Does the `|` separate the conditions that are added with `&`? Or does the first `&` separate them, as in the first condition must be met _and_ one of the next two conditions, and the last one? It's unclear to me how these conditions are combined exactly... Also, why are there parentheses around every single condition?

Reply:
I've sorted it! 
it's your version, but with & and | instead of && and || since if you put two they work only for the first element in list, and if you put one they process the all vector. Hence: 
signif.outer <- ((dataframe$ci.low > 0 && dataframe$ci.up > 0) ||
                 (dataframe$ci.low < 0 && dataframe$ci.up < 0))

Reply:
Great! Good to hear!